src/wallet/mod.rs

```diff
@@ -366,67 +332,236 @@ where
                 false => 0,
                 true => max_addresses,
             };
 
-            if self
-                .database
-                .borrow()
-                .get_script_pubkey_from_path(KeychainKind::Internal, max_address.saturating_sub(1))?
-                .is_none()
-            {
-                debug!("caching internal addresses");
-                new_addresses_cached = true;
-                self.cache_addresses(KeychainKind::Internal, 0, max_address)?;
+            let next_to_derive = self.next_derivation_index(keychain);
+            if next_to_derive > end {
+                return false;
+            }
+
+            for index in next_to_derive..=end {
+                let spk = descriptor
+                    .at_derivation_index(index)
+                    .derived_descriptor(&self.secp)
+                    .expect("the descritpor cannot need hardened derivation")
+                    .script_pubkey();
+
+                self.spk_tracker.add_spk((keychain, index), spk);
             }
         }
-        Ok(new_addresses_cached)
+        true
+    }
+
+    /// Derive all addresses up to and including `index`.
+    pub fn ensure_derived_up_to(&mut self, keychain: KeychainKind, index: u32) -> bool {
+        let descriptor = self.get_descriptor_for_keychain(keychain).clone();
+        let end = match descriptor.has_wildcard() {
+            false => 0,
+            true => index,
+        };
+
+        let next_to_derive = self.next_derivation_index(keychain);
+        if next_to_derive > end {
+            return false;
+        }
+
+        for index in next_to_derive..=end {
+            let spk = descriptor
+                .at_derivation_index(index)
+                .derived_descriptor(&self.secp)
+                .expect("the descritpor cannot need hardened derivation")
+                .script_pubkey();
+
+            self.spk_tracker.add_spk((keychain, index), spk);
+        }
+
+        true
     }
 
     /// Return whether or not a `script` is part of this wallet (either internal or external)
-    pub fn is_mine(&self, script: &Script) -> Result<bool, Error> {
-        self.database.borrow().is_mine(script)
+    pub fn is_mine(&self, script: &Script) -> bool {
+        self.spk_tracker.index_of_spk(script).is_some()
     }
 
     /// Return the list of unspent outputs of this wallet
     ///
     /// Note that this method only operates on the internal database, which first needs to be
     /// [`Wallet::sync`] manually.
-    pub fn list_unspent(&self) -> Result<Vec<LocalUtxo>, Error> {
-        Ok(self
-            .database
-            .borrow()
-            .iter_utxos()?
-            .into_iter()
-            .filter(|l| !l.is_spent)
-            .collect())
+    pub fn list_unspent(&self) -> Vec<LocalUtxo> {
+        self.spk_tracker
+            .iter_unspent(&self.sparse_chain, &self.tx_graph)
+            .map(|((keychain, _), txo)| LocalUtxo {
```

Comment:
This reminds me: the 1.0 release is a good opportunity to rename `LocalUtxo` to `LocalTxOut` :)

Reply:
Note i didn't do this here. I think it's possible we could get rid of this type in favor of using a `bdk_chain` one.